### PR TITLE
bluetooth: id: fix adv sets with same id use different RPA

### LINKED ISSUE #64625: Bluetooth: Host: advertising sets linked with the same identity may use different RPAs
**Describe the bug**

Advertising sets linked with the same Bluetooth identity may use different RPAs, which breaks the assumption introduced by the following PR:

https://github.com/zephyrproject-rtos/zephyr/pull/55449

This misbehaviour happens under certain conditions and when the developer uses the `rpa_expired` callback API:
```
/**
 * @brief The RPA validity of the advertising set has expired.
 *
 * This callback notifies the application that the RPA validity of
 * the advertising set has expired. The user can use this callback
 * to synchronize the advertising payload update with the RPA rotation.
 *
 * @param adv  The advertising set object.
 *
 * @return true to rotate the current RPA, or false to use it for the
 *         next rotation period.
 */
bool (*rpa_expired)(struct bt_le_ext_adv *adv);
```
Let's assume that the application defines two advertising sets that are linked to the same Bluetooth identity. Both sets start to advertise with the **RPA X**.  At some point, the RPA is timed out and the `rpa_expired` callback fires for both advertising sets. If the first advertising set chooses to withhold the RPA rotation by returning `false` in the `rpa_expired` callback and the second set progresses with the RPA rotation by returning `true` in the `rpa_expired` callback, then the first set will keep using the **RPA X** in the subsequent RPA rotation period and the second set will start to advertise with a new **RPA Y**. In this case, two advertising sets belonging to the same Bluetooth identity will use two different RPAs: **RPA X** and **RPA Y**.

**To Reproduce**

1. Configure the Bluetooth sample to use the Advertising Extension (`CONFIG_BT_EXT_ADV`) and Privacy (`CONFIG_BT_PRIVACY`). Make sure to set a proper Kconfig option to have the possibility to register at least two advertising sets (`CONFIG_BT_EXT_ADV_MAX_ADV_SET=2`).

   @alwa-nordic EDIT since https://github.com/zephyrproject-rtos/zephyr/pull/64733: Set `CONFIG_BT_RPA_SHARING=y`.
3. In the application code, create two advertising sets using the `bt_le_ext_adv_create` API and register the `rpa_expired` callback for both of them.
4. Set two different payloads using the `bt_le_ext_adv_set_data` API to easily distinguish two advertising sets on the scanning device.
5. Always return `false` In the `rpa_expired` callback implementation for the first advertising set.
6. Always return `true` In the `rpa_expired` callback implementation for the second advertising set.
7. Start the advertising for both sets using the `bt_le_ext_adv_start` API.
8. Build and flash the application.
9. See that both advertising sets use the same RPA during the first rotation.
10. Wait for the RPA rotation timeout.
11. See that the first advertising set uses the RPA from step 9 and the second advertising set uses a new one.

**Expected behavior**

After RPA timeout, both advertising sets should either stay with the same RPA or migrate to the new one based on the return value of the `rpa_expired` API. 

**Impact**

The feature for sharing the same RPA between the advertising sets does not work reliably.

**Environment (please complete the following information):**

 - OS: Linux
 - Toolchain: https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.16.3
 - Commit SHA or Version used: https://github.com/zephyrproject-rtos/zephyr/commit/67676d8fbcf5d742bdc6bbf0635d79499fa66062

### PR DESCRIPTION
**Pull request by niym-ot**
The fix is to check if any of the adv set's rpa expired callback returns false, then all adv sets continues with the old RPA.

Note: Fix is applicable only to adv sets which shares the same id.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/64625

### CODE REVIEW

**Comment by Thalley** [comment]
@niym-ot modified your PR text to use the proper way of linking PRs with issues (see https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)

**Comment by jori-nordic** [comment]
Could you add a bsim test for this functionality?

**Comment by alwa-nordic** [comment]
Can we add a new test instead of modifying an old one? I think it's much easier to navigate single-purpose tests with a clearly stated test purpose.

File: tests/bsim/bluetooth/host/privacy/peripheral/src/tester_rpa_rotation.c [hunk 1]
```diff
@@ -56,15 +56,16 @@ static void validate_rpa_addr_generated_for_adv_sets(void)
 			FAIL("RPA not same for adv sets with same id and RPA sharing enabled\n");
 		}
 	}
-	if (bt_addr_le_eq(&adv_set_data[0].old_addr, &adv_set_data[3].old_addr)) {
+	if (bt_addr_le_eq(&adv_set_data[0].old_addr, &adv_set_data[2].old_addr)) {
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/privacy/peripheral/src/tester_rpa_rotation.c:59]: Please move the refactors of other tests to a separate PR. You can tag it trivial.
You can keep the rename in this one.

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		switch (i)  {
+		case ADV_SET_INDEX_1:
+		case ADV_SET_INDEX_2:
+			create_adv(&adv_set[i], ID_1, true);
+			break;
+		case ADV_SET_INDEX_3:
+			create_adv(&adv_set[i], ID_2, true);
+			break;
+		case ADV_SET_INDEX_4:
+			create_adv(&adv_set[i], ID_2, false);
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c:152]: Could you post a follow-up PR where you add another id w/ two other adv sets, that return `false` for the first one and `true` for the second one? Ie started in reverse order so the iterator hits the `false` first.

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c (hunk 2 above)
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c:152]: clarify: this comment is not a blocker.

File: subsys/bluetooth/host/id.c [hunk 3]
```diff
@@ -198,23 +198,31 @@ int bt_id_set_adv_random_addr(struct bt_le_ext_adv *adv,
...
 #endif
-	if (rpa_invalid) {
-		atomic_clear_bit(adv->flags, BT_ADV_RPA_VALID);
 
-#if defined(CONFIG_BT_RPA_SHARING)
-		bt_addr_copy(&bt_dev.rpa[adv->id], BT_ADDR_NONE);
-#endif
+	if (IS_ENABLED(CONFIG_BT_RPA_SHARING)) {
+
+		if (adv->id >= bt_dev.id_count) {
```
[the remark this answers is not in the document]
Comment (niym-ot) [reply subsys/bluetooth/host/id.c:214]: This is boundary check

File: subsys/bluetooth/host/id.c [hunk 4]
```diff
@@ -198,23 +198,31 @@ int bt_id_set_adv_random_addr(struct bt_le_ext_adv *adv,
 	return 0;
 }
 
-static void adv_rpa_expired(struct bt_le_ext_adv *adv)
+static void adv_rpa_expired(struct bt_le_ext_adv *adv, void *data)
```
[the remark this answers is not in the document]
Comment (niym-ot) [reply subsys/bluetooth/host/id.c:205]: Done

File: tests/bsim/bluetooth/host/privacy/peripheral/src/tester_rpa_rotation.c (hunk 1 above)
Comment (niym-ot) [reply tests/bsim/bluetooth/host/privacy/peripheral/src/tester_rpa_rotation.c]: Removed the changes

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c (hunk 2 above)
Comment (niym-ot) [reply tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c:152]: Noted. Will create a pull request for the same.

File: subsys/bluetooth/host/id.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/bluetooth/host/id.c:205]: did you push? I don't see the changes

[the remark this answers is not in the document]
Comment (niym-ot) [reply subsys/bluetooth/host/id.c:205]: I added the comments above the adv_rpa_expired function and now added it in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/bluetooth.h#L164 and pushed the changes. Please check.

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c (hunk 2 above)
Comment (niym-ot) [reply tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_expired.c:152]: https://github.com/zephyrproject-rtos/zephyr/issues/69969

**Review by jhedberg — DISMISSED** [review]

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_rotation.c [hunk 5]
```diff
@@ -110,11 +111,10 @@ void start_advertising(void)
 
 	for (int i = 0; i < CONFIG_BT_EXT_ADV_MAX_ADV_SET; i++) {
 
-		if (i != ADV_SET_INDEX_THREE) {
-			/* Create advertising set 1 and 2 with same id */
+		/* Create first 2 advertising sets with one id and remaining with another id */
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_rotation.c:114]: please revert changes to this test

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_rotation.c (hunk 5 above)
Comment (niym-ot) [reply tests/bsim/bluetooth/host/privacy/peripheral/src/dut_rpa_rotation.c]: Done

**Review by jhedberg — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

**Pull request merged**